freshmaker/pyxis_gql.py

```diff
@@ -652,46 +649,26 @@ def find_images_by_name_version(
 
         ds = self.dsl_schema
-        page_num = 0
         # Pyxis is having an issue of returning images for such query criteria (ISV-3642),
         # so we use a much smaller page size in this particular query
-        page_size = 20
-
-        while True:
-            query_dsl = ds.Query.find_images(
-                page=page_num,
-                page_size=page_size,
-                # Sort by image NVR (brew.build), newer image should have higher NVR
-                sort_by=[{"field": "brew.build", "order": "DESC"}],
-                filter=query_filter,
-            ).select(
-                ds.ContainerImagePaginatedResponse.error.select(
-                    ds.ResponseError.status,
-                    ds.ResponseError.detail,
-                ),
-                ds.ContainerImagePaginatedResponse.page,
-                ds.ContainerImagePaginatedResponse.page_size,
-                ds.ContainerImagePaginatedResponse.total,
-                ds.ContainerImagePaginatedResponse.data.select(*self._get_image_projection()),
-            )
 
-            result = self.query(query_dsl)
-            data = result["find_images"]["data"]
-            # Data is empty when there are no more results
-            if not data:
-                break
-            images.extend(data)
-
-            if len(images) >= limit:
-                break
-
-            # If page_size >= total, means all results have been fetched in the first page
-            if result["find_images"]["page_size"] >= result["find_images"]["total"]:
-                break
-            page_num += 1
+        query_dsl = ds.Query.find_images(
+            page=0,
+            # "page_size" is equal to 4, because the maximum arches supported
+            # by RH containers is 4 (amd64, ppc64le, aarch64 and s390x)
+            page_size=4,
+            # Sort by image NVR (brew.build), newer image should have higher NVR
+            sort_by=[{"field": "brew.build", "order": "DESC"}],
+            filter=query_filter,
+        ).select(
+            ds.ContainerImagePaginatedResponse.error.select(
+                ds.ResponseError.status,
+                ds.ResponseError.detail,
+            ),
+            ds.ContainerImagePaginatedResponse.page,
+            ds.ContainerImagePaginatedResponse.page_size,
+            ds.ContainerImagePaginatedResponse.total,
+            ds.ContainerImagePaginatedResponse.data.select(*self._get_image_projection()),
+        )
 
-        # Note: this can truncate the list with keeping some arches of an image build but
-        # removing other arches, but we don't care about it in our use cases, we just want
-        # the most higher NVR builds, if you're calling this with a much smaller limit
-        # number (e.g.: 1, 2, 3) or for different purpose, you should pay attention to this.
-        return images[:limit]
+        return self.query(query_dsl)["find_images"]["data"]
```

Comment:
This works when image supports 4 arches, but when it supports just 1, 2 or 3 arches, this will return images for more than one NVR. For example, when it supports 3 arches, this will return 2 NVRs like:
```
foo-125
foo-125
foo-125
foo-125
```

And I checked with Pyxis, seems sorting by `brew.build` doesn't work for us, because Pyxis considers `foo-99` is higher than `foo-125`. 

I also tried sorting by `repositories.published_date`, in some cases it doesn't work either, for example, the query for `s2i-base-container-1-*` images returns this:
```
s2i-base-container-1-465.1698060557.amd64
s2i-base-container-1-465.1697652944.amd64
s2i-base-container-1-465.1698060557.s390x
s2i-base-container-1-465.1698060557.ppc64le
s2i-base-container-1-465.1697652944.ppc64le
s2i-base-container-1-465.1698060557.arm64
s2i-base-container-1-465.1697652944.arm64
s2i-base-container-1-465.1697652944.s390x
...
```
so probably we should still query with more page_size (e.g. 20), but sort by `repositories.published_date`, then compare the NVRs in result in freshmaker and only return the images for the highest NVR.
```
latest_nvr = images[0]["brew"]["build"]

for img in images[1:]:
    parsed_latest_nvr = kobo.rpmlib.parse_nvr(latest_nvr)

    current_nvr = img["brew"]["build"]
    parsed_current_nvr = kobo.rpmlib.parse_nvr(current_nvr)
    if kobo.rpmlib.compare_nvr(parsed_current_nvr, parsed_latest_nvr) == 1:
        latest_nvr = current_nvr

return [img for img in images if img["brew"]["build"] == latest_nvr]
```

Reply:
Gotcha. Thanks for the catch. I'll make the change and I will try query pyxis too, perhaps adding some test with a real query result data.